Update Executor interface for Run and Exec

coryb opened the pull request:
First step for #749 refactor Executor interface to allow for Run (pid1) vs Exec'ing into running containers.  Implements proposal outlined in: https://github.com/moby/buildkit/issues/749#issuecomment-655206042

tonistiigi commented on executor/executor.go:
```diff
@@ -28,9 +28,15 @@ type Mount struct {
+type ProcessInfo struct {
+	Meta           Meta
+	Stdin          io.ReadCloser
+	Stdout, Stderr io.WriteCloser
```
I guess TODO in here is `Resize <-chan Size`. Can be next PR

tonistiigi commented on executor/runcexecutor/executor.go:
```diff
@@ -310,6 +315,62 @@ func (w *runcExecutor) Exec(ctx context.Context, meta executor.Meta, root cache.
 }
 
+func (w *runcExecutor) Exec(ctx context.Context, id string, process executor.ProcessInfo) error {
+	// first verify the container is running, if we get an error assume the container
```
Could we do some sync with go mutex/cond/channel in here first. Eg. detect that `Run()` has been called for the ID and it has not returned yet. I guess there is still a race when `runc` binary is being called that needs to be checked with retries but first check and reacting to pid1 returning would be fast then. Also, maybe`Run()` should take a channel that is closed to indicate started event? Ideally, when pid1 returns while this method is running all the errors returned in here would be about pid1 closing, and not for example that open fails because state file has been deleted.

coryb replied:
yeah, I struggled a bit with race conditions between pid1 and exec.  I think a channel passed to Run seems reasonable.  Since runc.Runc is blocking I can setup a short goroutine to poll for "running" state before sending on the channel to avoid the ~200ms delay between runc.Run being called and the container being running.  I will try to rework it a bit to avoid the Exec retry loop.

tonistiigi replied:
Something like

```
Run(id string, started chan<- struct{}) {
  done = make(chan error, 1)
  
  e.mu.Lock()
  e.m[id] = done
  defer {
    e.mu.Lock()
    delete(e.m, id)
    e.mu.Unlock()
    done <- err
  }()
  e.mu.Unlock()
  
  close(started)
}

Exec(id string) {
  e.mu.Lock()
  done, ok := e.m[id]
  if !ok {
    return error404
  }
  e.mu.Unlock()
  
  defer func() {
    select {
    case err := <-done:
      err = errors.Wrapf(err, "pid1 returned with error")
    default:
    }
  }()
}
``` 
?

I still think retry is needed as well because of the way we call runc without getting the create event.

coryb replied:
makes sense, I will update.

tonistiigi commented on executor/runcexecutor/executor.go:
```diff
@@ -310,6 +315,62 @@ func (w *runcExecutor) Exec(ctx context.Context, meta executor.Meta, root cache.
+	}
+	if len(process.Meta.Env) > 0 {
+		// merge exec env with pid1 env
+		spec.Process.Env = append(spec.Process.Env, process.Meta.Env...)
```
I don't think we want to merge.

coryb replied:
just to confirm ... if PATH is set on pid1, and on a subsequent Exec we set just TERM, we expect the Exec to run without any PATH set?

tonistiigi replied:
`PATH` is kind-of special, eg. we have https://github.com/moby/buildkit/blob/master/solver/llbsolver/ops/exec.go#L728 and should probably do the same for exec. Do you see a case where exec would not know the env of the pid1 in case it wants to reuse some of its env and can't just resend it?

coryb replied:
No, that seems fine to me. I was just not sure if we expected the clients to resend the entire env on exec.

tonistiigi replied:
It seems safer this way to let the client have the full control when they don't want the same env.

coryb commented on worker/runc/runc_test.go:
```diff
@@ -202,11 +205,71 @@ func TestRuncWorkerNoProcessSandbox(t *testing.T) {
 	require.Equal(t, string(selfCmdline), stdout.String())
 }
 
+func TestRuncWorkerExec(t *testing.T) {
```
okay, will look into sharing some testing code to dedup.

tonistiigi replied:
Just the test(s) itself. Don't try to dedupe the setup logic.

coryb commented:
@tonistiigi thanks for the help with the review.  I think I have addressed all your issues, hopefully I captured what you wanted.

Looks like one test failed:
https://travis-ci.org/github/moby/buildkit/jobs/707044698#L2818

Not sure if this could be related to my PR or if the tests are sometimes flaky.

tonistiigi commented on executor/containerdexecutor/executor.go:
```diff
@@ -45,33 +48,56 @@ func New(client *containerd.Client, root, cgroup string, networkProviders map[pb
 	resolvConf, err := oci.GetResolvConf(ctx, w.root, nil, w.dnsConfig)
 	if err != nil {
-		return err
+		return sendErr(done, err)
```
This is better if wrapped in a single `defer` so there is no risk that wrapping was forgotten.

coryb replied:
Sure I can do that. I tend to avoid that pattern myself since I have always found the unintentional redefinition of `err` to be problematic and harder to find, ie:
```
func Exec(...) (err error) {
  defer func() {
    done <- err
  }()
  resolvConf, err := oci.GetResolvConf(ctx, w.root, nil, w.dnsConfig)
  if err != nil {
    return err
}
```
would send `nil` on `done` if `GetResolveConf` fails.

tonistiigi replied:
> would send nil on done if GetResolveConf fails.

No, it does not :)

coryb replied:
Hah, whoa, TIL.  I tried it on play just to sanity check, I must have misinterpreted the results, so much for sanity :(
I redid the test and of course you are correct, it works fine: https://play.golang.org/p/TvbDMw6q5Z2 I would have sworn I ran into this issue a while ago, maybe older Go versions had this issue?  Dunno, anyway, yeah that will make the change much easier :)

tonistiigi replied:
maybe this is what you were thinking https://play.golang.org/p/YYOCf97H0il

coryb replied:
I have no idea, maybe I ran into something like you suggest.  I tried on an old go1.2 runtime, it was fine there also.  So I guess I have just been wrong for years, o'well, glad you set me straight :)

tonistiigi commented on executor/runcexecutor/executor.go:
```diff
@@ -119,18 +123,39 @@ func New(opt Opt, networkProviders map[pb.NetMode]network.Provider) (executor.Ex
 		dns:              opt.DNS,
 		oomScoreAdj:      opt.OOMScoreAdj,
+		running:          make(map[string]chan error),
+		mu:               sync.Mutex{},
```
Don't need to initialize

tonistiigi commented on executor/containerdexecutor/executor.go:
```diff
@@ -31,6 +32,8 @@ type containerdExecutor struct {
 	cgroupParent     string
 	dnsConfig        *oci.DNSConfig
+	running          map[string]chan error
+	mu               *sync.Mutex
```
Don't see why this needs to be a pointer

coryb replied:
Yeah, it was b/c we have `func (w containerdExecutor) Run` instead of taking a pointer type, so that would result in copies of the sync.Mutex.  I can just change the function to accept a `*containerdExecutor` instead.

tonistiigi replied:
Yes, `*containerdExecutor` is better

tonistiigi commented on executor/runcexecutor/executor.go:
```diff
@@ -279,37 +307,111 @@ func (w *runcExecutor) Exec(ctx context.Context, meta executor.Meta, root cache.
+
+	state, _ := w.runc.State(ctx, id)
+	for {
+		if state != nil && state.Status == "running" {
```
I think we should recheck `w.running` in here after timeout and return out if not found. Basically move the "running" check inside the `for{}`.

tonistiigi commented on worker/tests/common.go:
```diff
@@ -0,0 +1,201 @@
+		}, started)
+	})
+
+	<-started
```
add timeout condition with `select`

tonistiigi commented on worker/tests/common.go:
```diff
@@ -0,0 +1,201 @@
+		Meta: executor.Meta{
+			Args: []string{"sh", "-c", "cat > /tmp/msg"},
+		},
+		Stdin:  &nopReadCloser{stdin},
```
nopcloser is in stdlib https://golang.org/pkg/io/ioutil/#NopCloser

coryb commented:
I think I addressed all the issues for round2, let me know if you find more issues.

hinshun commented:
@tonistiigi Yes, I'm out on a trip atm, but I'll give this a review tmrw morning.

hinshun commented on executor/containerdexecutor/executor.go:
```diff
@@ -209,5 +245,89 @@ func (w containerdExecutor) Exec(ctx context.Context, meta executor.Meta, root c
+
+	proc := spec.Process
+
+	// TODO how do we get rootfsPath for oci.GetUser in case user passed in username rather than uid:gid?
```
Do we need to fix this now? It looks like there is some existing code that deals with errors returned from `oci.ParseUIDGID(meta.User)` here: https://github.com/moby/buildkit/blob/5909d1642e8ece08e7aa64c95103c8fcffd2feb9/executor/containerdexecutor/executor.go#L107-L114

coryb replied:
Yeah, the problem is that we need the rootfsPath to call oci.GetUser, which I could not figure out how to access via containerd apis.  It was available when creating the container, but I could not figure out how to determine it after creation.  We could track it in the same map where we track the `done` channel, then I could use it later at Exec time?

hinshun commented on executor/executor.go:
```diff
@@ -28,9 +28,15 @@ type Mount struct {
 type Executor interface {
-	// TODO: add stdout/err
-	Exec(ctx context.Context, meta Meta, rootfs cache.Mountable, mounts []Mount, stdin io.ReadCloser, stdout, stderr io.WriteCloser) error
+	Run(ctx context.Context, id string, rootfs cache.Mountable, mounts []Mount, process ProcessInfo, started chan<- struct{}) error
```
Can we add some godocs and details about the usage of `started`?

hinshun commented:
Only change is [godocs](https://github.com/moby/buildkit/pull/1563/commits/caac9457888435ee0cda97b89a142196cd4acd5b), so I'll go ahead and merge this now.
